### PR TITLE
feat: support `Bytes` type return from contract

### LINKED ISSUE #723: Add support for the new stdlib type `Bytes`
The PR for this type can be found here: https://github.com/FuelLabs/sway/pull/3454

Related: https://github.com/FuelLabs/fuels-rs/issues/709

### PR DESCRIPTION
**Pull request by iqdecay**
This PR closes #723.
EDIT: needs #883 to go in first.

### CODE REVIEW

File: packages/fuels-types/src/param_types.rs [hunk 1]
```diff
@@ -68,38 +70,62 @@ impl ParamType {
         }
     }
 
-    pub fn contains_no_nested_vectors(&self) -> bool {
+    pub fn calculate_num_of_elements(&self, bytes: &[u8]) -> Result<usize> {
+        let memory_size = match self {
+            // The `Bytes` type in the VM uses byte-packing, don't multiply by WORD_SIZE
+            ParamType::Bytes => Ok(ParamType::U8.compute_encoding_width()),
```
Comment (segfault-magnet) [inline packages/fuels-types/src/param_types.rs:76]: I don't like the idea that `U8.compute_encoding_width()` returns 1 WORD (as described in the fn doc) meaning 8 B and we're reinterpreting it as 1 B by deciding not to multiply by `WORD_SIZE`. I'd rather we do something like `std::mem::size_of::<u8>()` which does return the number of bytes and does not confuse the meaning of `compute_encoding_width`.

File: packages/fuels-core/src/abi_decoder.rs [hunk 2]
```diff
@@ -65,6 +65,7 @@ impl ABIDecoder {
             ParamType::U64 => Self::decode_u64(bytes),
             ParamType::Bool => Self::decode_bool(bytes),
             ParamType::Byte => Self::decode_byte(bytes),
```
[the remark this answers is not in the document]
Comment (iqdecay) [reply packages/fuels-core/src/abi_decoder.rs:67]: Ok so it seems that #867 is unwarranted then?

[the remark this answers is not in the document]
Comment (segfault-magnet) [reply packages/fuels-core/src/abi_decoder.rs:67]: I'd close it. I couldn't get Sway to recognize `byte` as a type anymore. Dug around the compiler, doesn't seem to be there either. Finally found the [commit](https://github.com/FuelLabs/sway/commit/2e324774ae1f9e356f8f3fdcd410b18cd7727bab) that removed it, back in October '22.

File: packages/fuels-code-gen/src/program_bindings/resolved_type.rs [hunk 3]
```diff
@@ -99,6 +101,39 @@ pub(crate) fn resolve_type(
...
+        Some(ResolvedType {
+            type_name,
+            generic_params: vec![],
+        })
+    } else {
+        None
+    }
+}
+
+fn to_raw_untyped_ptr(
```
[the remark this answers is not in the document]
Comment (iqdecay) [reply packages/fuels-code-gen/src/program_bindings/resolved_type.rs:121]: nice  I didn't think of that!

File: packages/fuels-core/src/abi_encoder.rs [hunk 4]
```diff
@@ -123,7 +123,19 @@ impl ABIEncoder {
...
             Token::RawSlice(_) => {
-                unimplemented!("Encoding a raw_slice is currently not supported by the Fuel spec.")
+                unimplemented!(
+                    "Encoding a `RawSlice` type is currently not supported by the Fuel spec."
+                )
+            }
+            Token::Bytes(_) => {
+                unimplemented!(
+                    "Encoding a `Bytes` type is currently not supported by the Fuel spec."
+                )
```
[the remark this answers is not in the document]
Comment (iqdecay) [reply packages/fuels-core/src/abi_encoder.rs:133]: Will keep this in mind for later!

File: packages/fuels-types/src/param_types.rs (hunk 1 above)
Comment (iqdecay) [reply packages/fuels-types/src/param_types.rs:76]: Agreed!

File: packages/fuels-programs/src/call_utils.rs [hunk 5]
```diff
@@ -263,22 +269,32 @@ pub(crate) fn get_single_call_instructions(
...
-    if let ParamType::Vector(inner_param_type) = output_param_type {
-        let inner_type_byte_size: u16 =
-            (inner_param_type.compute_encoding_width() * WORD_SIZE) as u16;
+    if output_param_type.is_vm_heap_type() {
+        let inner_type_byte_size = match output_param_type {
+            ParamType::Vector(inner_param_type) => {
+                Ok((inner_param_type.compute_encoding_width() * WORD_SIZE) as u16)
+            }
+            // `Bytes` type is byte-packed in the VM, so don't multiply by WORD_SIZE
+            ParamType::Bytes => Ok(ParamType::Byte.compute_encoding_width() as u16),
```
[the remark this answers is not in the document]
Comment (iqdecay) [reply packages/fuels-programs/src/call_utils.rs:278]: Tracking removing `Byte` in #880

File: packages/fuels-core/src/abi_decoder.rs [hunk 6]
```diff
@@ -73,11 +74,39 @@ impl ABIDecoder {
...
+            .map_err(|e| error!(InvalidData, "{e}"))?
+            .into_iter()
+            .map(u8::from)
+            .collect();
+
+        Ok(DecodeResult {
+            token: Token::Bytes(u8_vec),
+            bytes_read,
+        })
+    }
```
[the remark this answers is not in the document]
Comment (iqdecay) [reply packages/fuels-core/src/abi_decoder.rs:106]: ty

File: packages/fuels-programs/src/call_utils.rs [hunk 7]
```diff
@@ -263,22 +269,32 @@ pub(crate) fn get_single_call_instructions(
...
+            _ => Err(error!(
```
[the remark this answers is not in the document]
Comment (iqdecay) [reply packages/fuels-programs/src/call_utils.rs:279]: very elegant

File: packages/fuels/tests/types/bytes/Forc.toml [hunk 8]
```diff
@@ -0,0 +1,7 @@
+[project]
+authors = ["lulwat"]
```
Comment (MujkicA) [inline packages/fuels/tests/types/bytes/Forc.toml:2]: lul
```suggestion
authors = ["Fuel Labs <contact@fuel.sh>"]
```

File: packages/fuels-test-helpers/src/signers.rs [hunk 9]
```diff
@@ -139,6 +139,7 @@ mod tests {
         Ok(())
     }
 
+    #[ignore]
```
Comment (MujkicA) [inline packages/fuels-test-helpers/src/signers.rs:142]: I assume this was just for local testing

**Review by MujkicA — COMMENTED** [review]
Looks good

**Review by Salka1988 — APPROVED** [review]
:+1:

File: packages/fuels-types/src/param_types.rs [hunk 10]
```diff
@@ -1295,53 +1334,116 @@ mod tests {
...
             variants: EnumVariants::new(tuples_no_nested_vec)?,
             generics: param_types_nested_vec
         }
-        .contains_nested_vectors());
+        .contains_nested_heap_types());
+        Ok(())
+    }
 
+    #[test]
+    fn contains_no_nested_heap_type_nested_bytes_complex_types() -> Result<()> {
```
Comment (segfault-magnet) [inline packages/fuels-types/src/param_types.rs:1387]: This naming is confusing. Kind of seems to have two parts. `contains_no_nested_heap_type` and `nested_bytes_complex_types`. But 'complex' types containing `Bytes` **have** nested heap types.

File: packages/fuels-types/src/param_types.rs [hunk 11]
```diff
@@ -67,42 +68,75 @@ impl ParamType {
...
         }
     }
 
-    fn uses_vectors(&self) -> bool {
+    fn uses_heap_types(&self) -> bool {
         match &self {
             ParamType::Vector(..) => true,
-            ParamType::Array(param_type, ..) => param_type.uses_vectors(),
-            ParamType::Tuple(param_types, ..) => Self::any_nested_vectors(param_types),
+            ParamType::Bytes => true,
```
Comment (segfault-magnet) [inline packages/fuels-types/src/param_types.rs:101]: You can merge these into one pattern:
```rust
            ParamType::Vector(..) | ParamType::Bytes => true,
```

File: packages/fuels-types/src/param_types.rs [hunk 12]
```diff
@@ -67,42 +68,75 @@ impl ParamType {
...
         param_types
             .into_iter()
-            .any(|param_type| param_type.uses_vectors())
+            .any(|param_type| param_type.uses_heap_types())
     }
 
-    pub fn is_vector(&self) -> bool {
-        matches!(self, Self::Vector(..))
+    pub fn is_vm_heap_type(&self) -> bool {
+        matches!(self, ParamType::Vector(..)) || matches!(self, ParamType::Bytes)
```
Comment (segfault-magnet) [inline packages/fuels-types/src/param_types.rs:127]: Can be replaced by one `matches!`:
```suggestion
        matches!(self, ParamType::Vector(..) | ParamType::Bytes)
```

File: packages/fuels-types/src/param_types.rs [hunk 13]
```diff
@@ -1295,53 +1334,116 @@ mod tests {
...
+            variants: EnumVariants::new(tuples_with_nested_bytes)?,
+            generics: param_types_no_nested_bytes
+        }
+        .contains_nested_heap_types());
+        assert!(ParamType::Enum {
+            name: "EnumName".to_string(),
+            variants: EnumVariants::new(tuples_no_nested_bytes)?,
+            generics: param_types_nested_bytes
+        }
+        .contains_nested_heap_types());
```
Comment (segfault-magnet) [inline packages/fuels-types/src/param_types.rs:1446]: You can use a helper fn to remove some of the noise around calling `assert!` and `contains_nested_heap_types` all the time:
```suggestion
        let is_nested = |param_type: ParamType| assert!(param_type.contains_nested_heap_types());
        let not_nested = |param_type: ParamType| assert!(!param_type.contains_nested_heap_types());

        not_nested(base_bytes.clone());
        is_nested(ParamType::Vector(Box::from(base_bytes.clone())));

        not_nested(ParamType::Array(Box::from(ParamType::U8), 10));
        is_nested(ParamType::Array(Box::from(base_bytes), 10));

        not_nested(ParamType::Tuple(param_types_no_nested_bytes.clone()));
        is_nested(ParamType::Tuple(param_types_nested_bytes.clone()));

        let not_nested_struct = ParamType::Struct {
            name: "StructName".to_string(),
            generics: param_types_no_nested_bytes.clone(),
            fields: tuples_no_nested_bytes.clone(),
        };
        not_nested(not_nested_struct);

        let nested_struct = ParamType::Struct {
            name: "StructName".to_string(),
            generics: param_types_nested_bytes.clone(),
            fields: tuples_no_nested_bytes.clone(),
        };
        is_nested(nested_struct);

        let nested_struct = ParamType::Struct {
            name: "StructName".to_string(),
            generics: param_types_no_nested_bytes.clone(),
            fields: tuples_with_nested_bytes.clone(),
        };
        is_nested(nested_struct);

        let not_nested_enum = ParamType::Enum {
            name: "EnumName".to_string(),
            variants: EnumVariants::new(tuples_no_nested_bytes.clone())?,
            generics: param_types_no_nested_bytes.clone(),
        };
        not_nested(not_nested_enum);

        let nested_enum = ParamType::Enum {
            name: "EnumName".to_string(),
            variants: EnumVariants::new(tuples_with_nested_bytes)?,
            generics: param_types_no_nested_bytes,
        };
        is_nested(nested_enum);

        let nested_enum = ParamType::Enum {
            name: "EnumName".to_string(),
            variants: EnumVariants::new(tuples_no_nested_bytes)?,
            generics: param_types_nested_bytes,
        };
        is_nested(nested_enum);
```

File: packages/fuels-types/src/param_types.rs [hunk 14]
```diff
@@ -1295,53 +1334,116 @@ mod tests {
...
 
-        assert!(!ParamType::Array(Box::from(ParamType::U8), 10).contains_nested_vectors());
-        assert!(ParamType::Array(Box::from(base_vector), 10).contains_nested_vectors());
+        assert!(!ParamType::Array(Box::from(ParamType::U8), 10).contains_nested_heap_types());
+        assert!(ParamType::Array(Box::from(base_vector), 10).contains_nested_heap_types());
 
-        assert!(!ParamType::Tuple(param_types_no_nested_vec.clone()).contains_nested_vectors());
-        assert!(ParamType::Tuple(param_types_nested_vec.clone()).contains_nested_vectors());
+        assert!(!ParamType::Tuple(param_types_no_nested_vec.clone()).contains_nested_heap_types());
+        assert!(ParamType::Tuple(param_types_nested_vec.clone()).contains_nested_heap_types());
```
Comment (segfault-magnet) [inline packages/fuels-types/src/param_types.rs:1344]: This test might also benefit from the helper suggested in the other test.

File: packages/fuels-core/src/abi_decoder.rs [hunk 15]
```diff
@@ -73,11 +73,19 @@ impl ABIDecoder {
...
+    fn decode_bytes(bytes: &[u8]) -> Result<DecodeResult> {
+        Ok(DecodeResult {
+            token: Token::Bytes(bytes.to_vec()),
+            bytes_read: bytes.len(),
+        })
+    }
     fn decode_vector(param_type: &ParamType, bytes: &[u8]) -> Result<DecodeResult> {
-        let num_of_elements = calculate_num_of_elements(param_type, bytes)?;
+        let num_of_elements =
+            ParamType::Vector(Box::from(param_type.clone())).calculate_num_of_elements(bytes)?;
```
[the remark this answers is not in the document]
Comment (iqdecay) [reply packages/fuels-core/src/abi_decoder.rs:87]: I like this!

**Review by hal3e — APPROVED** [review]
Well done!

**Pull request merged**